The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27547279882/job/81424197090) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/81424197090) [(plain)](https://github.com/rust-lang/rust/commit/7fd6d53d2e7d4cac27aa5698facc4bd2584e3eff/checks/81424197090/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6912 files
tidy check
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision [unspecified] should not specify `needs-llvm-components:` as it doesn't need `--target`
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision TRAP should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mips-mti-none-elf should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mips-unknown-linux-gnu should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mips-unknown-linux-musl should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mips-unknown-linux-uclibc should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mips64-openwrt-linux-musl should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mips64-unknown-linux-gnuabi64 should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mips64-unknown-linux-muslabi64 should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mips64el-unknown-linux-gnuabi64 should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mips64el-unknown-linux-muslabi64 should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsel-mti-none-elf should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsel-sony-psp should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsel-sony-psx should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsel-unknown-linux-gnu should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsel-unknown-linux-musl should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsel-unknown-linux-uclibc should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsel-unknown-netbsd should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsel-unknown-none should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsisa32r6-unknown-linux-gnu should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsisa32r6el-unknown-linux-gnu should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsisa64r6-unknown-linux-gnuabi64 should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: /checkout/tests/assembly-llvm/mips-div-no-trap.rs: revision mipsisa64r6el-unknown-linux-gnuabi64 should specify `needs-llvm-components: mips` as it has `--target` set
tidy [target-specific-tests (tests)]: FAIL
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
---
info: ✓ ES-Check passed! All files are ES10 compatible.
typechecking javascript files
tidy: The following check failed: target-specific-tests (tests)
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --ci=true --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1624:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1511:29

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1511:29
   5: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Tidy>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1596:36
   6: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1433:21
   7: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   8: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
   9: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1139:9
  10: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1118:14
  11: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  12: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  13: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:250:5
  14: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/sys/backtrace.rs:166:18
  15: std::rt::lang_start::<()>::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:206:18
  16: <&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:287:21
  17: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
  18: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:544:19
  19: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panic.rs:359:14
  20: std::rt::lang_start_internal::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:175:24
  21: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
---
  28: __libc_start_main
  29: _start


Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:02:43
  local time: Mon Jun 15 12:55:32 UTC 2026
  network time: Mon, 15 Jun 2026 12:55:33 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

--- a/compiler/rustc_target/src/spec/targets/mips64_openwrt_linux_musl.rs
+++ b/compiler/rustc_target/src/spec/targets/mips64_openwrt_linux_musl.rs
@@ -2,7 +2,7 @@
 
 use rustc_abi::Endian;
 
-use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     let mut base = base::linux_musl::opts();
@@ -28,6 +28,7 @@ pub(crate) fn target() -> Target {
             endian: Endian::Big,
             mcount: "_mcount".into(),
             llvm_abiname: LlvmAbi::N64,
+            llvm_args: cvs!["-mno-check-zero-division"],
             ..base
         },
     }

--- a/compiler/rustc_target/src/spec/targets/mips64_unknown_linux_gnuabi64.rs
+++ b/compiler/rustc_target/src/spec/targets/mips64_unknown_linux_gnuabi64.rs
@@ -1,6 +1,6 @@
 use rustc_abi::Endian;
 
-use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -23,6 +23,7 @@ pub(crate) fn target() -> Target {
             max_atomic_width: Some(64),
             mcount: "_mcount".into(),
             llvm_abiname: LlvmAbi::N64,
+            llvm_args: cvs!["-mno-check-zero-division"],
 
             ..base::linux_gnu::opts()
         },

--- a/compiler/rustc_target/src/spec/targets/mips64_unknown_linux_muslabi64.rs
+++ b/compiler/rustc_target/src/spec/targets/mips64_unknown_linux_muslabi64.rs
@@ -1,6 +1,6 @@
 use rustc_abi::Endian;
 
-use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     let mut base = base::linux_musl::opts();
@@ -24,6 +24,7 @@ pub(crate) fn target() -> Target {
             endian: Endian::Big,
             mcount: "_mcount".into(),
             llvm_abiname: LlvmAbi::N64,
+            llvm_args: cvs!["-mno-check-zero-division"],
             ..base
         },
     }

--- a/compiler/rustc_target/src/spec/targets/mips64el_unknown_linux_gnuabi64.rs
+++ b/compiler/rustc_target/src/spec/targets/mips64el_unknown_linux_gnuabi64.rs
@@ -1,4 +1,4 @@
-use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -20,6 +20,7 @@ pub(crate) fn target() -> Target {
             max_atomic_width: Some(64),
             mcount: "_mcount".into(),
             llvm_abiname: LlvmAbi::N64,
+            llvm_args: cvs!["-mno-check-zero-division"],
 
             ..base::linux_gnu::opts()
         },

--- a/compiler/rustc_target/src/spec/targets/mips64el_unknown_linux_muslabi64.rs
+++ b/compiler/rustc_target/src/spec/targets/mips64el_unknown_linux_muslabi64.rs
@@ -1,4 +1,4 @@
-use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     let mut base = base::linux_musl::opts();
@@ -21,6 +21,7 @@ pub(crate) fn target() -> Target {
             cfg_abi: CfgAbi::Abi64,
             mcount: "_mcount".into(),
             llvm_abiname: LlvmAbi::N64,
+            llvm_args: cvs!["-mno-check-zero-division"],
             ..base
         },
     }

--- a/compiler/rustc_target/src/spec/targets/mips_mti_none_elf.rs
+++ b/compiler/rustc_target/src/spec/targets/mips_mti_none_elf.rs
@@ -2,7 +2,7 @@ use rustc_abi::Endian;
 
 use crate::spec::{
     Arch, Cc, LinkerFlavor, Lld, LlvmAbi, PanicStrategy, RelocModel, Target, TargetMetadata,
-    TargetOptions,
+    TargetOptions, cvs,
 };
 
 pub(crate) fn target() -> Target {
@@ -26,6 +26,7 @@ pub(crate) fn target() -> Target {
             cpu: "mips32r2".into(),
 
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             max_atomic_width: Some(32),
 
             features: "+mips32r2,+soft-float,+noabicalls".into(),

--- a/compiler/rustc_target/src/spec/targets/mips_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/mips_unknown_linux_gnu.rs
@@ -1,6 +1,6 @@
 use rustc_abi::Endian;
 
-use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -19,6 +19,7 @@ pub(crate) fn target() -> Target {
             cpu: "mips32r2".into(),
             features: "+mips32r2,+fpxx,+nooddspreg".into(),
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             max_atomic_width: Some(32),
             mcount: "_mcount".into(),
 

--- a/compiler/rustc_target/src/spec/targets/mips_unknown_linux_musl.rs
+++ b/compiler/rustc_target/src/spec/targets/mips_unknown_linux_musl.rs
@@ -1,6 +1,6 @@
 use rustc_abi::Endian;
 
-use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     let mut base = base::linux_musl::opts();
@@ -21,6 +21,7 @@ pub(crate) fn target() -> Target {
         options: TargetOptions {
             endian: Endian::Big,
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             mcount: "_mcount".into(),
             ..base
         },

--- a/compiler/rustc_target/src/spec/targets/mips_unknown_linux_uclibc.rs
+++ b/compiler/rustc_target/src/spec/targets/mips_unknown_linux_uclibc.rs
@@ -1,6 +1,6 @@
 use rustc_abi::Endian;
 
-use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -19,6 +19,7 @@ pub(crate) fn target() -> Target {
             cpu: "mips32r2".into(),
             features: "+mips32r2,+soft-float".into(),
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             max_atomic_width: Some(32),
             mcount: "_mcount".into(),
 

--- a/compiler/rustc_target/src/spec/targets/mipsel_mti_none_elf.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsel_mti_none_elf.rs
@@ -2,7 +2,7 @@ use rustc_abi::Endian;
 
 use crate::spec::{
     Arch, Cc, LinkerFlavor, Lld, LlvmAbi, PanicStrategy, RelocModel, Target, TargetMetadata,
-    TargetOptions,
+    TargetOptions, cvs,
 };
 
 pub(crate) fn target() -> Target {
@@ -26,6 +26,7 @@ pub(crate) fn target() -> Target {
             cpu: "mips32r2".into(),
 
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             max_atomic_width: Some(32),
 
             features: "+mips32r2,+soft-float,+noabicalls".into(),

--- a/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_gnu.rs
@@ -1,4 +1,4 @@
-use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -17,6 +17,7 @@ pub(crate) fn target() -> Target {
             cpu: "mips32r2".into(),
             features: "+mips32r2,+fpxx,+nooddspreg".into(),
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             max_atomic_width: Some(32),
             mcount: "_mcount".into(),
 

--- a/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_musl.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_musl.rs
@@ -1,4 +1,4 @@
-use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     let mut base = base::linux_musl::opts();
@@ -16,6 +16,11 @@ pub(crate) fn target() -> Target {
         pointer_width: 32,
         data_layout: "e-m:m-p:32:32-i8:8:32-i16:16:32-i64:64-n32-S64".into(),
         arch: Arch::Mips,
-        options: TargetOptions { llvm_abiname: LlvmAbi::O32, mcount: "_mcount".into(), ..base },
+        options: TargetOptions {
+            llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
+            mcount: "_mcount".into(),
+            ..base
+        },
     }
 }

--- a/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_uclibc.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_uclibc.rs
@@ -1,4 +1,4 @@
-use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -17,6 +17,7 @@ pub(crate) fn target() -> Target {
             cpu: "mips32r2".into(),
             features: "+mips32r2,+soft-float".into(),
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             max_atomic_width: Some(32),
             mcount: "_mcount".into(),
 

--- a/compiler/rustc_target/src/spec/targets/mipsel_unknown_netbsd.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsel_unknown_netbsd.rs
@@ -1,6 +1,6 @@
 use rustc_abi::Endian;
 
-use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     let mut base = base::netbsd::opts();
@@ -21,6 +21,7 @@ pub(crate) fn target() -> Target {
         options: TargetOptions {
             features: "+soft-float".into(),
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             mcount: "__mcount".into(),
             endian: Endian::Little,
             ..base

--- a/compiler/rustc_target/src/spec/targets/mipsel_unknown_none.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsel_unknown_none.rs
@@ -4,7 +4,7 @@
 
 use crate::spec::{
     Arch, Cc, LinkerFlavor, Lld, LlvmAbi, PanicStrategy, RelocModel, Target, TargetMetadata,
-    TargetOptions,
+    TargetOptions, cvs,
 };
 
 pub(crate) fn target() -> Target {
@@ -25,6 +25,7 @@ pub(crate) fn target() -> Target {
             cpu: "mips32r2".into(),
             features: "+mips32r2,+soft-float,+noabicalls".into(),
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             max_atomic_width: Some(32),
             linker: Some("rust-lld".into()),
             panic_strategy: PanicStrategy::Abort,

--- a/compiler/rustc_target/src/spec/targets/mipsisa32r6_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsisa32r6_unknown_linux_gnu.rs
@@ -1,6 +1,6 @@
 use rustc_abi::Endian;
 
-use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -19,6 +19,7 @@ pub(crate) fn target() -> Target {
             cpu: "mips32r6".into(),
             features: "+mips32r6".into(),
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             max_atomic_width: Some(32),
             mcount: "_mcount".into(),
 

--- a/compiler/rustc_target/src/spec/targets/mipsisa32r6el_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsisa32r6el_unknown_linux_gnu.rs
@@ -1,4 +1,4 @@
-use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -17,6 +17,7 @@ pub(crate) fn target() -> Target {
             cpu: "mips32r6".into(),
             features: "+mips32r6".into(),
             llvm_abiname: LlvmAbi::O32,
+            llvm_args: cvs!["-mno-check-zero-division"],
             max_atomic_width: Some(32),
             mcount: "_mcount".into(),
 

--- a/compiler/rustc_target/src/spec/targets/mipsisa64r6_unknown_linux_gnuabi64.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsisa64r6_unknown_linux_gnuabi64.rs
@@ -1,6 +1,6 @@
 use rustc_abi::Endian;
 
-use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -23,6 +23,7 @@ pub(crate) fn target() -> Target {
             max_atomic_width: Some(64),
             mcount: "_mcount".into(),
             llvm_abiname: LlvmAbi::N64,
+            llvm_args: cvs!["-mno-check-zero-division"],
 
             ..base::linux_gnu::opts()
         },

--- a/compiler/rustc_target/src/spec/targets/mipsisa64r6el_unknown_linux_gnuabi64.rs
+++ b/compiler/rustc_target/src/spec/targets/mipsisa64r6el_unknown_linux_gnuabi64.rs
@@ -1,4 +1,4 @@
-use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base};
+use crate::spec::{Arch, CfgAbi, LlvmAbi, Target, TargetMetadata, TargetOptions, base, cvs};
 
 pub(crate) fn target() -> Target {
     Target {
@@ -20,6 +20,7 @@ pub(crate) fn target() -> Target {
             max_atomic_width: Some(64),
             mcount: "_mcount".into(),
             llvm_abiname: LlvmAbi::N64,
+            llvm_args: cvs!["-mno-check-zero-division"],
 
             ..base::linux_gnu::opts()
         },

--- a/tests/assembly-llvm/mips-div-no-trap.rs
+++ b/tests/assembly-llvm/mips-div-no-trap.rs
@@ -1,0 +1,124 @@
+// Test that there's no conditional trap for zero divisor for all mips
+// targets by default.  Division by zero is defined as panic so the trap is
+// redundant.
+//
+//@ add-minicore
+//@ assembly-output: emit-asm
+//
+// See https://github.com/llvm/llvm-project/pull/204386.
+//@ compile-flags: -Copt-level=3
+//
+//@ revisions: mips64el-unknown-linux-gnuabi64
+//@[mips64el-unknown-linux-gnuabi64] compile-flags: --target=mips64el-unknown-linux-gnuabi64
+//@[mips64el-unknown-linux-gnuabi64] needs-llvm-components: mips
+//@[mips64el-unknown-linux-gnuabi64] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mips64el-unknown-linux-muslabi64
+//@[mips64el-unknown-linux-muslabi64] compile-flags: --target=mips64el-unknown-linux-muslabi64
+//@[mips64el-unknown-linux-muslabi64] needs-llvm-components: mips
+//@[mips64el-unknown-linux-muslabi64] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mips64-openwrt-linux-musl
+//@[mips64-openwrt-linux-musl] compile-flags: --target=mips64-openwrt-linux-musl
+//@[mips64-openwrt-linux-musl] needs-llvm-components: mips
+//@[mips64-openwrt-linux-musl] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mips64-unknown-linux-gnuabi64
+//@[mips64-unknown-linux-gnuabi64] compile-flags: --target=mips64-unknown-linux-gnuabi64
+//@[mips64-unknown-linux-gnuabi64] needs-llvm-components: mips
+//@[mips64-unknown-linux-gnuabi64] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mips64-unknown-linux-muslabi64
+//@[mips64-unknown-linux-muslabi64] compile-flags: --target=mips64-unknown-linux-muslabi64
+//@[mips64-unknown-linux-muslabi64] needs-llvm-components: mips
+//@[mips64-unknown-linux-muslabi64] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsel-mti-none-elf
+//@[mipsel-mti-none-elf] compile-flags: --target=mipsel-mti-none-elf
+//@[mipsel-mti-none-elf] needs-llvm-components: mips
+//@[mipsel-mti-none-elf] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsel-sony-psp
+//@[mipsel-sony-psp] compile-flags: --target=mipsel-sony-psp
+//@[mipsel-sony-psp] needs-llvm-components: mips
+//@[mipsel-sony-psp] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsel-sony-psx
+//@[mipsel-sony-psx] compile-flags: --target=mipsel-sony-psx
+//@[mipsel-sony-psx] needs-llvm-components: mips
+//@[mipsel-sony-psx] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsel-unknown-linux-gnu
+//@[mipsel-unknown-linux-gnu] compile-flags: --target=mipsel-unknown-linux-gnu
+//@[mipsel-unknown-linux-gnu] needs-llvm-components: mips
+//@[mipsel-unknown-linux-gnu] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsel-unknown-linux-musl
+//@[mipsel-unknown-linux-musl] compile-flags: --target=mipsel-unknown-linux-musl
+//@[mipsel-unknown-linux-musl] needs-llvm-components: mips
+//@[mipsel-unknown-linux-musl] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsel-unknown-linux-uclibc
+//@[mipsel-unknown-linux-uclibc] compile-flags: --target=mipsel-unknown-linux-uclibc
+//@[mipsel-unknown-linux-uclibc] needs-llvm-components: mips
+//@[mipsel-unknown-linux-uclibc] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsel-unknown-netbsd
+//@[mipsel-unknown-netbsd] compile-flags: --target=mipsel-unknown-netbsd
+//@[mipsel-unknown-netbsd] needs-llvm-components: mips
+//@[mipsel-unknown-netbsd] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsel-unknown-none
+//@[mipsel-unknown-none] compile-flags: --target=mipsel-unknown-none
+//@[mipsel-unknown-none] needs-llvm-components: mips
+//@[mipsel-unknown-none] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsisa32r6el-unknown-linux-gnu
+//@[mipsisa32r6el-unknown-linux-gnu] compile-flags: --target=mipsisa32r6el-unknown-linux-gnu
+//@[mipsisa32r6el-unknown-linux-gnu] needs-llvm-components: mips
+//@[mipsisa32r6el-unknown-linux-gnu] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsisa32r6-unknown-linux-gnu
+//@[mipsisa32r6-unknown-linux-gnu] compile-flags: --target=mipsisa32r6-unknown-linux-gnu
+//@[mipsisa32r6-unknown-linux-gnu] needs-llvm-components: mips
+//@[mipsisa32r6-unknown-linux-gnu] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsisa64r6el-unknown-linux-gnuabi64
+//@[mipsisa64r6el-unknown-linux-gnuabi64] compile-flags: --target=mipsisa64r6el-unknown-linux-gnuabi64
+//@[mipsisa64r6el-unknown-linux-gnuabi64] needs-llvm-components: mips
+//@[mipsisa64r6el-unknown-linux-gnuabi64] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mipsisa64r6-unknown-linux-gnuabi64
+//@[mipsisa64r6-unknown-linux-gnuabi64] compile-flags: --target=mipsisa64r6-unknown-linux-gnuabi64
+//@[mipsisa64r6-unknown-linux-gnuabi64] needs-llvm-components: mips
+//@[mipsisa64r6-unknown-linux-gnuabi64] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mips-mti-none-elf
+//@[mips-mti-none-elf] compile-flags: --target=mips-mti-none-elf
+//@[mips-mti-none-elf] needs-llvm-components: mips
+//@[mips-mti-none-elf] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mips-unknown-linux-gnu
+//@[mips-unknown-linux-gnu] compile-flags: --target=mips-unknown-linux-gnu
+//@[mips-unknown-linux-gnu] needs-llvm-components: mips
+//@[mips-unknown-linux-gnu] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mips-unknown-linux-musl
+//@[mips-unknown-linux-musl] compile-flags: --target=mips-unknown-linux-musl
+//@[mips-unknown-linux-musl] needs-llvm-components: mips
+//@[mips-unknown-linux-musl] filecheck-flags: --check-prefix NOTRAP
+//@ revisions: mips-unknown-linux-uclibc
+//@[mips-unknown-linux-uclibc] compile-flags: --target=mips-unknown-linux-uclibc
+//@[mips-unknown-linux-uclibc] needs-llvm-components: mips
+//@[mips-unknown-linux-uclibc] filecheck-flags: --check-prefix NOTRAP
+//
+//@ revisions: TRAP
+//@[TRAP] compile-flags: --target=mips64el-unknown-linux-gnuabi64 -C llvm-args=-mno-check-zero-division=0
+//@[TRAP] needs-llvm-components: mips
+
+#![crate_type = "lib"]
+#![feature(no_core, intrinsics)]
+#![no_core]
+
+extern crate minicore;
+
+#[rustc_intrinsic]
+pub unsafe fn unchecked_div<T>(x: T, y: T) -> T;
+
+#[rustc_intrinsic]
+pub fn abort() -> !;
+
+// NOTRAP-NOT: teq
+// TRAP: teq
+#[no_mangle]
+pub fn div_i32(a: i32, b: i32) -> i32 {
+    match a {
+        0 => abort(),
+        -1 => match b {
+            -2147483648 => abort(),
+            _ => unsafe { unchecked_div(a, b) },
+        },
+        _ => unsafe { unchecked_div(a, b) },
+    }
+}

